### PR TITLE
[Distributed] Remove mangled names from API, move to RemoteCallTarget and pretty print it

### PR DESCRIPTION
**Pull request by ktoso**
This does a number of things based on review rounds:

- remove mangled names from any API at all, even tho it was just a name and we could have accepted otehr things, the name was bad
- move to accepting `RemoteCallTarget`
- remove `mangledName` on and rather provide an unmangled representation if we're able to

TODO: needs more "bad case" tests.

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by ktoso** [comment]
Unblocked by https://github.com/apple/swift-syntax/pull/371/files

File: stdlib/public/runtime/Casting.cpp [hunk 1]
```diff
@@ -255,6 +255,82 @@ swift::swift_getMangledTypeName(const Metadata *type) {
...
+      }
+    }
+  }
+
+  if (funcIdent) {
+    demangled += ")";
+  }
+
+  // We have to copy the string to be able to refer to it;
+  // TODO(distributed): currently we need to alloc/decalloc this each time, we should cache it instead
```
[the remark this answers is not in the document]
Comment (ktoso) [reply stdlib/public/runtime/Casting.cpp]: done!

File: stdlib/public/Distributed/DistributedActorSystem.swift [hunk 2]
```diff
@@ -172,23 +172,18 @@ extension DistributedActorSystem {
...
   public func executeDistributedTarget<Act, ResultHandler>(
     on actor: Act,
-    mangledTargetName: String,
+    target: RemoteCallTarget,
     invocationDecoder: inout InvocationDecoder,
     handler: ResultHandler
   ) async throws where Act: DistributedActor,
                        // Act.ID == ActorID, // FIXME(distributed): can we bring this back?
                        ResultHandler: DistributedTargetInvocationResultHandler {
-    // NOTE: this implementation is not the most efficient, nor final, version of this func
```
[the remark this answers is not in the document]
Comment (ktoso) [reply stdlib/public/Distributed/DistributedActorSystem.swift:181]: Hm, I guess there is room for optimization here still, I'll leave a note 👍

File: stdlib/public/Distributed/DistributedActorSystem.swift [hunk 3]
```diff
@@ -326,6 +323,52 @@ extension DistributedActorSystem {
...
+/// and pass it along to the remote system for in their `remoteCall`
+/// implementation. Alternative approaches are possible, where the identifiers
+/// are either compressed, cached, or represented in other ways, as long as the
+/// recipient system is able to determine which target was intended to be
+/// invoked.
+///
+/// The string representation will attempt to pretty print the target identifier,
+/// however its exact format is not specified and may change in future versions.
+@available(SwiftStdlib 5.7, *)
+public struct RemoteCallTarget: CustomStringConvertible {
```
Comment (ktoso) [inline stdlib/public/Distributed/DistributedActorSystem.swift:340]: I was quite torn on the representation of this to be honest.

We could internally store:

```
  private enum _Storage {
     case mangledName(String)
   }
```

but then the initializer would be mangledName and cause people to over worry about it again...

Internal representation we can always change in the future right @DougGregor ? The struct is not frozen.

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Pull request merged**